---
title: "RDA_Contaminants~Microbioma"
author: "Elena Baños Lara"
date: "29/4/2022"
output:
  html_document:
    theme: readable
    toc: yes
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::include_graphics
options(width = 50)
```

# Opening a new project:
## Necessary libraries:
```{r, message=FALSE, warning=FALSE}
 library(vegan)
# library(scales)
# library(RColorBrewer)
# library(VennDiagram)
 library(gplots)
 library(ggplot2)
# library(ggpubr)
# library(car)
# library(pairwiseAdonis)
# library(ggVennDiagram)
# library(ggplot2)
# library(eulerr)
 library(devtools)
# library(wrapr)
# library(tidyr)
 library(dplyr)
 library(stringr)
library(tidyr)
 library(reshape)
 library(reshape2)
# library(data.table)
# devtools::install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
# library(pairwiseAdonis)
# library(tibble)
# library(tidyverse)
# library(dtplyr)
# install.packages("rlang")
# library(rlang)
library("clusterSim")
#library("BiodiversityR")
#library("ggrepel")
#library("ggforce")
#library("data.table")
library(ggrepel)

#devtools::install_github("gavinsimpson/ggvegan")
```

## RDA
RDA: combines regression and PCA, it is an extension of regression analysis to model multivariate response data. RDA computes axes that are linear combinations of the explanatory variables (in order of which explain the most variation of the species matrix). The axes are orthogonal to eachother (i.e. right angles). It is constrained because you are directly testing the influence of explanatory variables.
```{r}
# Fes el dataset bo
setwd("E:/styela/Molecular/Microbioma/microbioma")

microbioma <-read.table("asv_april2020_relativefreq.txt", header = TRUE, sep = "\t")
#microbioma <- microbioma[c(1:19409),]
#microbioma <- t(microbioma)
microbioma <- as.data.frame(microbioma)

pollutans <- read.table("pollutants_rda.txt", header = TRUE, sep = "\t")
pollutans <- pollutans[c(2:55),]
ASVlist <- as.vector(pollutans$ID)
ASVlistt <- as.vector(pollutans$ID[pollutans$TISSUE!="2GILL"])
ASVlista <- as.vector(pollutans$ID[pollutans$STAGE!="1JUV"])
as.data.frame(pollutans)
metas <- pollutans[,c(1,11:13)]
meta <- pollutans[,c(11:13)]
pollutants <- pollutans[,c(2:10)]
pollutants <- data.Normalization(pollutants, type = "n1", normalization = "column")
#samples <-read.table("asv_taxonomy_april2020_rda.txt", header = TRUE, sep = "\t")

```

####RDA general

```{r}

samples <-microbioma[rownames(microbioma) %in% ASVlist,]

write.table(samples, "asv_taxonomy_april2020_rda.txt", sep = "\t")  ####TREURE MANUALMENT ELS ESPAIS!! REEMPLAZAR

sty.rda <- rda(samples~.,data=pollutants)

summary(sty.rda)##AQUI ES VEUEN ELS % de variabilitat explicats pel RDA Buscar: accumulated constrained eigenvalues, els necessitarem per despr?s


summary(eigenvals(sty.rda, model = "constrained")) #es veuen els % de cada eix
pdf("pollutants_eigenvalues_RDA.pdf")
screeplot(sty.rda) # els % de cada eix gràfic
dev.off()

####treure tests per veure si l'RDA és significatiu en general i per cada eix####
#signif.full <- anova.cca(sty.rda, parallel=getOption("mc.cores")) # default is permutation=999 PER veure si el model es significatiu
#write.table(signif.full,file="pollutants_model_anova_fullaxes_RDA.txt", sep="\t")
#signif.axis <- anova.cca(sty.rda, by="axis", parallel=getOption("mc.cores")) # PER veure si cada eix del model es significatiu
#write.table(signif.axis,file="pollutants_model_anova_eachaxis_RDA.txt", sep="\t")
vif.cca(sty.rda) #


####Plotejar els RDA per individu####
rdaplot<-summary(sty.rda)
arrows <- as.data.frame(rdaplot$biplot) #treure la info de les fletxes
#write.table(arrows,file="pollutants_envs_asso_RDA.txt", sep="\t")
rdaplot <- as.data.frame(rdaplot$sites) # treure la info dels indvs
rdaplot<- cbind(rdaplot, meta) # assignar pop a individu

head(rdaplot)

colnames(rdaplot) <- c("RDA1","RDA2","RDA3","RDA4","RDA5","RDA6","Pop","Comp","Stage") # canviar nom a columnes RDA
#row.names(arrows) <- c("Mean_Temp","Mean_Sal","Range_Temp","Range_sal","si","O2","CO2") #canviar nom a files de variables env

mycol <- c("#F4D03F","#EC7063","#85C1E9")                     #TUNIC,GILL,WATER
##colors del teu plot per cada pop

#plots correlatius dels primers 4 eixos
pdf("pollutants_mic_RDA_1vs2.pdf", width = 7, height = 6)
ggplot(data = rdaplot, aes(x = RDA1, y = RDA2))+
  geom_point(data = rdaplot, size = 4, aes(color=Stage, fill = Comp, shape=Pop,     alpha=1), stroke = 1.5)+
  geom_segment(data = arrows, aes(x = 0, xend = 0.5*RDA1, y = 0, yend = 0.5*RDA2),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) +    geom_vline(xintercept = 0, lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrows, aes(x= 0.55*RDA1, y = 0.55*RDA2, label = rownames(arrows)), 
            size = 5, hjust = 0.5)+
  #geom_text(data = prdav, aes(x = RDA1, y = RDA2, label = rownames(prdav)), 
  #         size = 2.5, col = "black", hjust = 1.2)+
  theme( legend.position="none", 
         panel.grid.minor = element_blank(), 
         panel.grid.major = element_blank(),
         panel.background = element_rect(fill = "transparent", colour = NA))+
  scale_fill_manual(values=mycol, labels=c("Tunic","Gill","Water"))+
  scale_shape_manual(values=c(25,23,21)) +
  scale_color_manual(values=c("#DB7093","#815c97","steelblue")) +
  theme_classic()+
  labs(x = paste("RDA1 (", round(summary(sty.rda)$cont$importance[2,1]*100, 2), "%)", sep = ""), y = paste("RDA2 (", round(summary(sty.rda)$cont$importance[2,2]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))   
dev.off()


pdf("pollutants_mic_RDA_2vs3.pdf", width = 7, height = 6)
ggplot(data = rdaplot, aes(x = RDA2, y = RDA3))+
  geom_point(data = rdaplot, size = 4, aes(color=Stage, fill = Comp, shape=Pop,     alpha=1), stroke = 1.5)+
  geom_segment(data = arrows, aes(x = 0, xend = 0.5*RDA2, y = 0, yend = 0.5*RDA3),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) +    geom_vline(xintercept = 0, lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrows, aes(x= 0.55*RDA2, y = 0.55*RDA3, label = rownames(arrows)), 
            size = 5, hjust = 0.5)+
  #geom_text(data = prdav, aes(x = RDA1, y = RDA2, label = rownames(prdav)), 
  #         size = 2.5, col = "black", hjust = 1.2)+
  theme( legend.position="none", 
         panel.grid.minor = element_blank(), 
         panel.grid.major = element_blank(),
         panel.background = element_rect(fill = "transparent", colour = NA))+
  scale_fill_manual(values=mycol, labels=c("Tunic","Gill","Water"))+
  scale_shape_manual(values=c(25,23,21)) +
  scale_color_manual(values=c("#DB7093","#815c97","steelblue")) +
  theme_classic()+
  labs(x = paste("RDA2 (", round(summary(sty.rda)$cont$importance[2,2]*100, 2), "%)", sep = ""), y = paste("RDA3 (", round(summary(sty.rda)$cont$importance[2,3]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))   
dev.off()
```

####RDA túniques

```{r}
samplest <-microbioma[rownames(microbioma) %in% ASVlistt,]
pollutantst <- pollutans[pollutans$ID %in% ASVlistt,]
pollutantst <- pollutantst[,c(2:10)]
metat <- metas[metas$ID %in% ASVlistt,]
metat <- metat[,c(2:4)]

sty.rdat <- rda(samplest ~ Al+As+V+Pb+B+Fe+Zn+Cu+Se ,data=pollutantst)

summary(sty.rdat)##AQUI ES VEUEN ELS % de variabilitat explicats pel RDA Buscar: accumulated constrained eigenvalues, els necessitarem per despr?s


summary(eigenvals(sty.rdat, model = "constrained")) #es veuen els % de cada eix
pdf("pollutantst_eigenvalues_RDA.pdf")
screeplot(sty.rdat) # els % de cada eix gràfic
dev.off()

####treure tests per veure si l'RDA és significatiu en general i per cada eix####
#signif.full <- anova.cca(sty.rdat, parallel=getOption("mc.cores")) # default is permutation=999 PER veure si el model es significatiu
#write.table(signif.full,file="pollutantst_model_anova_fullaxes_RDA.txt", sep="\t")
#signif.axis <- anova.cca(sty.rdat, by="axis", parallel=getOption("mc.cores")) # PER veure si cada eix del model es significatiu
#write.table(signif.axis,file="pollutantst_model_anova_eachaxis_RDA.txt", sep="\t")
vif.cca(sty.rdat) #


####Plotejar els RDA per individu####
rdaplott<-summary(sty.rdat)
arrowst <- as.data.frame(rdaplott$biplot) #treure la info de les fletxes
#write.table(arrowst,file="pollutantst_envs_asso_RDA.txt", sep="\t")
rdaplott <- as.data.frame(rdaplott$sites) # treure la info dels indvs
rdaplott<- cbind(rdaplott, metat) # assignar pop a individu

head(rdaplott)

colnames(rdaplott) <- c("RDA1","RDA2","RDA3","RDA4","RDA5","RDA6","Pop","Comp","Stage") # canviar nom a columnes RDA
#row.names(arrows) <- c("Mean_Temp","Mean_Sal","Range_Temp","Range_sal","si","O2","CO2") #canviar nom a files de variables env

mycol <- c("#F4D03F","#85C1E9")                     #TUNIC,GILL,WATER
##colors del teu plot per cada pop

#plots correlatius dels primers 4 eixos
pdf("pollutantst_mic_RDA_1vs2.pdf", width = 7, height = 6)
ggplot(data = rdaplott, aes(x = RDA1, y = RDA2))+
  geom_point(data = rdaplott, size = 4, aes(color=Stage, fill = Comp, shape=Pop,     alpha=1), stroke = 1.5)+
  geom_segment(data = arrowst, aes(x = 0, xend = 0.5*RDA1, y = 0, yend = 0.5*RDA2),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) +    geom_vline(xintercept = 0, lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowst, aes(x= 0.55*RDA1, y = 0.55*RDA2, label = rownames(arrowst)), 
            size = 5, hjust = 0.5)+
  #geom_text(data = prdav, aes(x = RDA1, y = RDA2, label = rownames(prdav)), 
  #         size = 2.5, col = "black", hjust = 1.2)+
  theme( legend.position="none", 
         panel.grid.minor = element_blank(), 
         panel.grid.major = element_blank(),
         panel.background = element_rect(fill = "transparent", colour = NA))+
  scale_fill_manual(values=mycol, labels=c("Tunic","Gill","Water"))+
  scale_shape_manual(values=c(25,23,21)) +
  scale_color_manual(values=c("#DB7093","#815c97","steelblue")) +
  theme_classic()+
  labs(x = paste("RDA1 (", round(summary(sty.rdat)$cont$importance[2,1]*100, 2), "%)", sep = ""), y = paste("RDA2 (", round(summary(sty.rdat)$cont$importance[2,2]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))   
dev.off()


pdf("pollutantst_mic_RDA_2vs3.pdf", width = 7, height = 6)
ggplot(data = rdaplott, aes(x = RDA2, y = RDA3))+
  geom_point(data = rdaplott, size = 4, aes(color=Stage, fill = Comp, shape=Pop,     alpha=1), stroke = 1.5)+
  geom_segment(data = arrowst, aes(x = 0, xend = 0.5*RDA2, y = 0, yend = 0.5*RDA3),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) +    geom_vline(xintercept = 0, lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowst, aes(x= 0.55*RDA2, y = 0.55*RDA3, label = rownames(arrows)), 
            size = 5, hjust = 0.5)+
  #geom_text(data = prdav, aes(x = RDA1, y = RDA2, label = rownames(prdav)), 
  #         size = 2.5, col = "black", hjust = 1.2)+
  theme( legend.position="none", 
         panel.grid.minor = element_blank(), 
         panel.grid.major = element_blank(),
         panel.background = element_rect(fill = "transparent", colour = NA))+
  scale_fill_manual(values=mycol, labels=c("Tunic","Gill","Water"))+
  scale_shape_manual(values=c(25,23,21)) +
  scale_color_manual(values=c("#DB7093","#815c97","steelblue")) +
  theme_classic()+
  labs(x = paste("RDA2 (", round(summary(sty.rdat)$cont$importance[2,2]*100, 2), "%)", sep = ""), y = paste("RDA3 (", round(summary(sty.rdat)$cont$importance[2,3]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))   
dev.off()
```

####RDA adults

```{r}
samplesa <-microbioma[rownames(microbioma) %in% ASVlista,]
pollutantsa <- pollutans[pollutans$ID %in% ASVlista,]
pollutantsa <- pollutantsa[,c(2:10)]
metaa <- metas[metas$ID %in% ASVlista,]
metaa <- metaa[,c(2:4)]

sty.rdaa <- rda(samplesa~.,data=pollutantsa)

summary(sty.rdaa)##AQUI ES VEUEN ELS % de variabilitat explicats pel RDA Buscar: accumulated constrained eigenvalues, els necessitarem per despr?s


summary(eigenvals(sty.rdaa, model = "constrained")) #es veuen els % de cada eix
pdf("pollutantsa_eigenvalues_RDA.pdf")
screeplot(sty.rdaa) # els % de cada eix gràfic
dev.off()

####treure tests per veure si l'RDA és significatiu en general i per cada eix####
#signif.full <- anova.cca(sty.rdaa, parallel=getOption("mc.cores")) # default is permutation=999 PER veure si el model es significatiu
#write.table(signif.full,file="pollutantsa_model_anova_fullaxes_RDA.txt", sep="\t")
#signif.axis <- anova.cca(sty.rdaa, by="axis", parallel=getOption("mc.cores")) # PER veure si cada eix del model es significatiu
#write.table(signif.axis,file="pollutantsa_model_anova_eachaxis_RDA.txt", sep="\t")
vif.cca(sty.rdaa) #


####Plotejar els RDA per individu####
rdaplota<-summary(sty.rdaa)
arrowsa <- as.data.frame(rdaplota$biplot) #treure la info de les fletxes
#write.table(arrowsa,file="pollutantsa_envs_asso_RDA.txt", sep="\t")
rdaplota <- as.data.frame(rdaplota$sites) # treure la info dels indvs
rdaplota<- cbind(rdaplota, metaa) # assignar pop a individu

head(rdaplota)

colnames(rdaplota) <- c("RDA1","RDA2","RDA3","RDA4","RDA5","RDA6","Pop","Comp","Stage") # canviar nom a columnes RDA
#row.names(arrows) <- c("Mean_Temp","Mean_Sal","Range_Temp","Range_sal","si","O2","CO2") #canviar nom a files de variables env

mycol <- c("#F4D03F","#EC7063","#85C1E9")                     #TUNIC,GILL,WATER
##colors del teu plot per cada pop

#plots correlatius dels primers 4 eixos
pdf("pollutantsa_mic_RDA_1vs2.pdf", width = 7, height = 6)
ggplot(data = rdaplota, aes(x = RDA1, y = RDA2))+
  geom_point(data = rdaplota, size = 4, aes(color=Stage, fill = Comp, shape=Pop,     alpha=1), stroke = 1.5)+
  geom_segment(data = arrowsa, aes(x = 0, xend = 0.5*RDA1, y = 0, yend = 0.5*RDA2),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) +    geom_vline(xintercept = 0, lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowsa, aes(x= 0.55*RDA1, y = 0.55*RDA2, label = rownames(arrowsa)), 
            size = 5, hjust = 0.5)+
  #geom_text(data = prdav, aes(x = RDA1, y = RDA2, label = rownames(prdav)), 
  #         size = 2.5, col = "black", hjust = 1.2)+
  theme( legend.position="none", 
         panel.grid.minor = element_blank(), 
         panel.grid.major = element_blank(),
         panel.background = element_rect(fill = "transparent", colour = NA))+
  scale_fill_manual(values=mycol, labels=c("Tunic","Gill","Water"))+
  scale_shape_manual(values=c(25,23,21)) +
  scale_color_manual(values=c("#815c97","steelblue")) +
  theme_classic()+
  labs(x = paste("RDA1 (", round(summary(sty.rdaa)$cont$importance[2,1]*100, 2), "%)", sep = ""), y = paste("RDA2 (", round(summary(sty.rdaa)$cont$importance[2,2]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))   
dev.off()


pdf("pollutantsa_mic_RDA_2vs3.pdf", width = 7, height = 6)
ggplot(data = rdaplota, aes(x = RDA2, y = RDA3))+
  geom_point(data = rdaplota, size = 4, aes(color=Stage, fill = Comp, shape=Pop,     alpha=1), stroke = 1.5)+
  geom_segment(data = arrowsa, aes(x = 0, xend = 0.5*RDA2, y = 0, yend = 0.5*RDA3),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) +    geom_vline(xintercept = 0, lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowsa, aes(x= 0.55*RDA2, y = 0.55*RDA3, label = rownames(arrowsa)), 
            size = 5, hjust = 0.5)+
  #geom_text(data = prdav, aes(x = RDA1, y = RDA2, label = rownames(prdav)), 
  #         size = 2.5, col = "black", hjust = 1.2)+
  theme( legend.position="none", 
         panel.grid.minor = element_blank(), 
         panel.grid.major = element_blank(),
         panel.background = element_rect(fill = "transparent", colour = NA))+
  scale_fill_manual(values=mycol, labels=c("Tunic","Gill","Water"))+
  scale_shape_manual(values=c(25,23,21)) +
  scale_color_manual(values=c("#815c97","steelblue")) +
  theme_classic()+
  labs(x = paste("RDA2 (", round(summary(sty.rdaa)$cont$importance[2,2]*100, 2), "%)", sep = ""), y = paste("RDA3 (", round(summary(sty.rdaa)$cont$importance[2,3]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))   
dev.off()
```

##Candidats de tunica

```{r}
load.rda <- scores(sty.rdat, choices=c(1:4), display="species")

hist(load.rda[,1], main="Loadings on RDA1") #cand de l'eix 1
hist(load.rda[,2], main="Loadings on RDA2") #cand de l'eix 2

outliers <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

cand1 <- outliers(load.rda[,1],3) # numero outliers a cada eix
cand2 <- outliers(load.rda[,2],3) # 


ncand <- length(cand1) + length(cand2) 
ncand

cand1 <- cbind.data.frame(rep(1,times=length(cand1)), names(cand1), unname(cand1)) #definir els SNPs a cada eix
cand2 <- cbind.data.frame(rep(2,times=length(cand2)), names(cand2), unname(cand2))

colnames(cand1) <- colnames(cand2) <- c("axis","ASV","loading")

cand <- rbind(cand1, cand2) #concatenar els SNPs candidats
cand$ASV <- as.character(cand$ASV)

foo <- matrix(nrow=(ncand), ncol=9)  # 8 columns for 8 predictors
colnames(foo) <- colnames(pollutantst)

samplest <- as.matrix(samplest)

for (i in 1:length(cand$ASV)) {
  nam <- cand[i,2]
  samplest.t <- samplest[,nam]
  foo[i,] <- apply(pollutantst,2,function(x) cor(x,samplest.t))
}

cand <- cbind.data.frame(cand,foo)  
head(cand)
cand <- cand[!duplicated(cand$ASV),] #elimina candidate duplicats

####dona a cada SNP candidat la seva variable amb la que es correlaciona i el seu grau de correlació####
for (i in 1:length(cand$ASV)) { #concatena a cada SNP les variables i la correlació amb aquest
  bar <- cand[i,]
  cand[i,13] <- names(which.max(abs(bar[4:10]))) # gives the variable
  cand[i,14] <- max(abs(bar[4:10]))              # gives the correlation
}

colnames(cand)[13] <- "predictor"
colnames(cand)[14] <- "correlation"

table(cand$predictor) 
write.table(cand,file="tunica_ont_Candidates_2eixos.txt", sep="\t")

sel <- cand$ASV
env <- cand$predictor
#env[env=="meanto"] <- 'orangered4'
 # env[env=="meanso"] <- '#1f78b4'
  #  env[env=="rangeto"] <- 'orangered1'
   #   env[env=="rangeso"] <- '#a6cee3'
    #    env[env=="si"] <- '#33a02c'
     #     env[env=="o2"] <- 'grey25'
      #      env[env=="spco2"] <- 'grey50'
          
          # color by predictor:
          col.pred <- rownames(sty.rdat$CCA$v) # pull the SNP names
          
          for (i in 1:length(sel)) {           # color code candidate SNPs
            foo <- match(sel[i],col.pred)
            col.pred[foo] <- env[i]
          }
          
          col.pred[grep("ASV",col.pred)] <- 'non-candidate' # non-candidate SNPs
            empty <- col.pred
            empty[grep("white",empty)] <- rgb(0,1,0, alpha=0) # transparent
            empty.outline <- ifelse(empty=="#00FF0000","#00FF0000","grey50")
            bg <- c('purple2','blue2',"grey80",'black','orange','red2') #c("Mean_sal","Mean_Temp","non-candidate","CO2","Range_Sal","Range_Temp","Si","O2"))+
            bg_col <- c('black','black',"NA",'black','black','black','black','black')
            
          colors <- as.data.frame(col.pred)
          col.pred <- as.vector(col.pred)
candplot <- summary(sty.rda) 
candplot <- as.data.frame(candplot$species)  
candplot<- cbind(candplot, colors)
head(candplot)
colnames(candplot) <- c("RDA1","RDA2","RDA3","RDA4","RDA5","RDA6","cand")

order <- c("As", "Pb", "B", "Zn", "V")

top <- candplot[grep("non-candidate", candplot$cand),]
top['ID'] <- ""
bottom <- candplot[-grep("non-candidate", candplot$cand),]
bottom <- bottom[order(match(bottom$cand,order)),]
bottom['ID'] <- rownames(bottom)
candplot <- rbind(top, bottom) 

tail(candplot)
head(candplot)

pdf("ASVt_pol_candidate_1vs2.pdf", width = 7, height = 6)     
ggplot(data = candplot, aes(x = RDA1, y = RDA2))+
  geom_point(data = candplot, pch = 21, size = 4, aes(color=candplot$cand, fill=candplot$cand, alpha=1))+
  geom_segment(data = arrowst, aes(x = 0, xend = 0.4*RDA1, y = 0, yend = 0.4*RDA2),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) + geom_vline(xintercept = 0, 
            lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowst, aes(x= 0.35*RDA1, y = 0.35*RDA2, label = rownames(arrowst)), 
            size = 5, hjust = 0.5)+
   geom_text_repel(data = candplot, aes(label = ID), 
            size = 2.5, hjust = 0.5)+
  scale_fill_manual(values=bg,labels=c("As", "B", "non-candidate","Pb", "V", "Zn"))+
  scale_color_manual(values=bg_col)+
  theme_classic()+
  labs(x = paste("RDA1 (", round(summary(sty.rdat)$cont$importance[2,1]*100, 2), "%)", sep = ""), 
       y = paste("RDA2 (", round(summary(sty.rdat)$cont$importance[2,2]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold")) 
dev.off()

pdf("ASVt_pol_candidate_3vs4.pdf", width = 7, height = 6)     
ggplot(data = candplot, aes(x = RDA3, y = RDA4))+
  geom_point(data = candplot, pch = 21, size = 4, aes(color=candplot$cand, fill=candplot$cand, alpha=1))+
  geom_segment(data = arrowst, aes(x = 0, xend = 0.3*RDA3, y = 0, yend = 0.3*RDA4),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) + geom_vline(xintercept = 0, 
              lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowst, aes(x= 0.33*RDA3, y = 0.33*RDA4, label = rownames(arrowst)), 
            size = 5, hjust = 0.5)+
  geom_text_repel(data = candplot, aes(label = ID), 
            size = 2.5, hjust = 0.5)+
  scale_fill_manual(values=bg,labels=c("As", "B", "non-candidate","Pb", "V", "Zn"))+
  scale_color_manual(values=bg_col)+
  theme_classic()+
  labs(x = paste("RDA3 (", round(summary(sty.rdat)$cont$importance[2,3]*100, 2), "%)", sep = ""), 
       y = paste("RDA4 (", round(summary(sty.rdat)$cont$importance[2,4]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))
dev.off()
```


##Candidats de teixit

```{r}
load.rdaa <- scores(sty.rdaa, choices=c(1:4), display="species")

hist(load.rdaa[,1], main="Loadings on RDA1") #cand de l'eix 1
hist(load.rdaa[,2], main="Loadings on RDA2") #cand de l'eix 2

outliers <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

cand1a <- outliers(load.rdaa[,1],3) # numero outliers a cada eix
cand2a <- outliers(load.rdaa[,2],3) # 


ncanda <- length(cand1a) + length(cand2a) 
ncanda

cand1a <- cbind.data.frame(rep(1,times=length(cand1a)), names(cand1a), unname(cand1a)) #definir els SNPs a cada eix
cand2a <- cbind.data.frame(rep(2,times=length(cand2a)), names(cand2a), unname(cand2a))

colnames(cand1a) <- colnames(cand2a) <- c("axis","ASV","loading")

canda <- rbind(cand1a, cand2a) #concatenar els SNPs candidats
canda$ASV <- as.character(canda$ASV)

fooa <- matrix(nrow=(ncanda), ncol=9)  # 8 columns for 8 predictors
colnames(fooa) <- colnames(pollutantsa)

samplesa <- as.matrix(samplesa)

for (i in 1:length(canda$ASV)) {
  nama <- canda[i,2]
  samplesa.a <- samplesa[,nama]
  fooa[i,] <- apply(pollutantsa,2,function(x) cor(x,samplesa.a))
}

canda <- cbind.data.frame(canda,fooa)  
head(canda)
canda <- canda[!duplicated(canda$ASV),] #elimina candidate duplicats

####dona a cada SNP candidat la seva variable amb la que es correlaciona i el seu grau de correlació####
for (i in 1:length(canda$ASV)) { #concatena a cada SNP les variables i la correlació amb aquest
  bara <- canda[i,]
  canda[i,13] <- names(which.max(abs(bara[4:10]))) # gives the variable
  canda[i,14] <- max(abs(bara[4:10]))              # gives the correlation
}

colnames(canda)[13] <- "predictor"
colnames(canda)[14] <- "correlation"

table(canda$predictor) 
write.table(canda,file="adults_teix_Candidates_2eixos.txt", sep="\t")

sela <- canda$ASV
enva <- canda$predictor
#env[env=="meanto"] <- 'orangered4'
 # env[env=="meanso"] <- '#1f78b4'
  #  env[env=="rangeto"] <- 'orangered1'
   #   env[env=="rangeso"] <- '#a6cee3'
    #    env[env=="si"] <- '#33a02c'
     #     env[env=="o2"] <- 'grey25'
      #      env[env=="spco2"] <- 'grey50'
          
          # color by predictor:
          col.preda <- rownames(sty.rdaa$CCA$v) # pull the SNP names
          
          for (i in 1:length(sel)) {           # color code candidate SNPs
            fooa <- match(sela[i],col.preda)
            col.preda[fooa] <- enva[i]
          }
          
          col.preda[grep("ASV",col.preda)] <- 'non-candidate' # non-candidate SNPs
            emptya <- col.preda
            emptya[grep("white",emptya)] <- rgb(0,1,0, alpha=0) # transparent
            empty.outlinea <- ifelse(empty=="#00FF0000","#00FF0000","grey50")
            bga <- c('purple2','blue2',"grey80",'black','orange') #c("Mean_sal","Mean_Temp","non-candidate","CO2","Range_Sal","Range_Temp","Si","O2"))+
            bg_cola <- c('black','black',"NA",'black','black','black')
            
          colorsa <- as.data.frame(col.preda)
          col.preda <- as.vector(col.preda)
candplota <- summary(sty.rdaa) 
candplota <- as.data.frame(candplota$species)  
candplota<- cbind(candplota, colorsa)
head(candplota)
colnames(candplota) <- c("RDA1","RDA2","RDA3","RDA4","RDA5","RDA6","cand")

ordera <- c("As", "B", "V", "Zn")

topa <- candplota[grep("non-candidate", candplota$cand),]
topa['ID'] <- ""
bottoma <- candplota[-grep("non-candidate", candplota$cand),]
bottoma <- bottoma[order(match(bottoma$cand,order)),]
bottoma['ID'] <- rownames(bottoma)
candplota <- rbind(topa, bottoma) 

tail(candplota)
head(candplota)

table(candplota$cand) 

pdf("ASVa_pol_candidate_1vs2.pdf", width = 7, height = 6)     
ggplot(data = candplota, aes(x = RDA1, y = RDA2))+
  geom_point(data = candplota, pch = 21, size = 4, aes(color=candplota$cand, fill=candplota$cand, alpha=1))+
  geom_segment(data = arrowsa, aes(x = 0, xend = 0.4*RDA1, y = 0, yend = 0.4*RDA2),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) + geom_vline(xintercept = 0, 
            lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowsa, aes(x= 0.35*RDA1, y = 0.35*RDA2, label = rownames(arrowsa)), 
            size = 5, hjust = 0.5)+
   geom_text_repel(data = candplota, aes(label = ID), 
            size = 2.5, hjust = 0.5)+
  scale_fill_manual(values=bga,labels=c("As", "B","non-candidate", "V", "Zn"))+
  scale_color_manual(values=bg_cola)+
  theme_classic()+
  labs(x = paste("RDA1 (", round(summary(sty.rdaa)$cont$importance[2,1]*100, 2), "%)", sep = ""), 
       y = paste("RDA2 (", round(summary(sty.rdaa)$cont$importance[2,2]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold")) 
dev.off()

pdf("ASVa_pol_candidate_3vs4.pdf", width = 7, height = 6)     
ggplot(data = candplota, aes(x = RDA3, y = RDA4))+
  geom_point(data = candplota, pch = 21, size = 4, aes(color=candplota$cand, fill=candplota$cand, alpha=1))+
  geom_segment(data = arrowsa, aes(x = 0, xend = 0.3*RDA3, y = 0, yend = 0.3*RDA4),
               arrow = arrow(length = unit(0.3, "cm")), colour = "steelblue", cex=1) +
  geom_hline(yintercept = 0, lty = "dotted", color="grey", cex=1) + geom_vline(xintercept = 0, 
              lty = "dotted", color="grey", cex=1) +
  geom_text(data = arrowsa, aes(x= 0.33*RDA3, y = 0.33*RDA4, label = rownames(arrowsa)), 
            size = 5, hjust = 0.5)+
  geom_text_repel(data = candplota, aes(label = ID), 
            size = 2.5, hjust = 0.5)+
  scale_fill_manual(values=bga,labels=c("As", "B","non-candidate", "V", "Zn"))+
  scale_color_manual(values=bg_cola)+
  theme_classic()+
  labs(x = paste("RDA3 (", round(summary(sty.rdaa)$cont$importance[2,3]*100, 2), "%)", sep = ""), 
       y = paste("RDA4 (", round(summary(sty.rdaa)$cont$importance[2,4]*100, 2), "%)", sep = ""))+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text = element_text(colour = "black", size = 12, face = "bold")) +
  theme(axis.title = element_text(size = 16, colour = "black", family = "Helvetica", face = "bold"))
dev.off()
```